core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java

```diff
@@ -573,6 +647,36 @@ private void assertClusterHealthStatus(RoutingAllocation allocation, ClusterHeal
         assertThat(clusterStateHealth.getStatus().ordinal(), lessThanOrEqualTo(expectedStatus.ordinal()));
     }
 
+    private AllocationDecider getNoDeciderThatAllowsForceAllocate() {
+        return new TestAllocateDecision(Decision.NO) {
+            @Override
+            public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                assert shardRouting.primary() : "cannot force allocate a non-primary shard " + shardRouting;
+                return Decision.YES;
+            }
+        };
+    }
+
+    private AllocationDecider getNoDeciderThatThrottlesForceAllocate() {
+        return new TestAllocateDecision(Decision.NO) {
+            @Override
+            public Decision canForceAllocatePrimary(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                assert shardRouting.primary() : "cannot force allocate a non-primary shard " + shardRouting;
+                return Decision.THROTTLE;
+            }
+        };
+    }
+
+    private AllocationDecider getNoDeciderThatDeniesForceAllocate() {
```

Comment:
we can have three methods here, but all of them could just share their code by calling a common method `getNoDeciderWithForceAllocate(Decision)`